gu-envman-api/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+extern crate actix;
+#[macro_use]
+extern crate serde_derive;
+extern crate actix_web;
+extern crate gu_net;
+extern crate serde;
+#[cfg(test)]
+extern crate serde_json;
+
+use actix::prelude::*;
+use gu_net::rpc::peer::PeerSessionInfo;
+use std::path::PathBuf;
+use std::{fmt, io};
+
+/// Errors
+// impl note: can not use error_chain bc it does not support SerDe
+#[derive(Serialize, Deserialize, Debug)]
+pub enum Error {
+    Error(String),
+    IoError(String),
+    NoSuchSession(String),
+    NoSuchChild(String),
+    UnknownEnv(String),
+}
+
+impl From<io::Error> for Error {
+    fn from(e: io::Error) -> Self {
+        Error::IoError(e.to_string())
+    }
+}
+
+impl From<actix::MailboxError> for Error {
+    fn from(e: MailboxError) -> Self {
+        Error::Error(format!("{}", e))
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Error::Error(msg) => write!(f, "error: {}", msg)?,
+            Error::IoError(msg) => write!(f, "IO error: {}", msg)?,
+            Error::NoSuchSession(msg) => write!(f, "session not found: {}", msg)?,
+            Error::NoSuchChild(msg) => write!(f, "child not found: {}", msg)?,
+            Error::UnknownEnv(env_id) => write!(f, "unknown exec environment: {}", env_id)?,
+        }
+        Ok(())
+    }
+}
+
+impl From<String> for Error {
+    fn from(msg: String) -> Self {
+        Error::Error(msg)
+    }
+}
+
+/// image with binaries and resources for given session
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct Image {
+    pub url: String,
+    pub hash: String,
+}
+
+/// Message for session creation: local provisioning: downloads and unpacks the binaries
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CreateSession {
+    pub env_type: String,
+    pub image: Image,
+    pub name: String,
+    pub tags: Vec<String>,
+    pub note: Option<String>,
+}
+
+impl CreateSession {
```

Comment:
shall we introduce some trait forcing ID definition for messages ie. like this

```suggestion
#![feature(associated_consts)]
trait OperationId {
    const ID: u32;
}

impl OperationId for CreateSession {
```

Reply:
1) the feature `associated_consts` has been stable since 1.20.0
2) this change has nothing to this pull request.

Reply:
ad1. so, we can use it
ad2. right, we can do it separately

Reply:
no attribute "feature" is only for unstable features. it will add compilation warning.